File: Assignment 1/bst_phonebook.py
class Node:
    # intialize values to left, right, name, phone and address nodes
    def __init__(self, name, phone, address):
        self.left = None
        self.right = None
        self.name = name
        self.phone = phone
        self.address = address
    # string function to build the strings and return an output with the variable names and their assigned values.
    def __str__(self):
        string = "" #empty
        string = string + f"Name: {self.name}\n"
        string = string + f"Phone Number: {self.phone}\n"
        string = string f"Address: {self.address}\n"
        return string


class Name:
    def __init__(self):
        # set the root to none
        self.root = None

    # str function to join the string message as one
    def __str__(self):
        node_strs = [str(node) for node in self.inorder()]
        return "-----------------\n".join(node_strs)
    
    # insert function to insert nodes into BST
    def insert(self, name, phone, address):
        if self.root is None:
            self.root = Node(name, phone, address)
        else:
            self.rec_insert(self.root, name, phone, address)

    # recursion of insert
    def rec_insert(self, presentNode, name, phone, address):
        if name < presentNode.name:
            if presentNode.left:
                self.rec_insert(presentNode.left, name, phone, address)
            else:
                presentNode.left = Node(name, phone, address)
        elif name > presentNode.name:
            if presentNode.right:
                self.rec_insert(presentNode.right, name, phone, address)
            else:
                presentNode.right = Node(name, phone, address)
    # find function to search for certain node within BST
    def find(self, name):
        return self.rec_find(name, self.root)
        if name is None:
            print(node)
        else:
            return False
    # recursion of find in order to find node.
    def rec_find(self, name, node):
        if name == node.name:
            # if it exists print success
            print(node)
            return True
        
        if name < node.name:
            if node.left == None:
                # if it does not exist print error
                print("Name does not exist in the Phonebook\n")
                return False
            return self.rec_find(name, node.left)
        
        if node.right == None:
            # if it does not exist print error
            print("Name does not exist in the Phonebook\n")
            return False
        return self.rec_find(name, node.right)
    '''
    remove function in order to remove a name from the phonebook
    function was created with the aid of https://www.youtube.com/watch?v=LSju119w8BE
    '''
    def remove(self, name):
        # initaialize root as empty
        if self.root is None:
            return False
        
        # search for name within tree

        elif self.root.name == name:
            if self.root.left is None and self.root.right is None:
                self.root == None
            elif self.root and self.root.right is None:
                self.root == self.root.left
            elif self.root.left is None and self.root.right:
                self.root == self.root.right
            elif self.root.left and self.root.right:
                deleteParent = self.root
                deleteNode = self.root.right
                while deleteNode.left:
                    deleteParent = deleteNode
                    deleteNode = deleteNode.left
                
                self.root.name = deleteNode.name
                self.root.phone = deleteNode.phone
                self.root.address = deleteNode.address

                if deleteNode.right:
                    if deleteParent.name > deleteNode.name:
                        deleteParent.left = deleteNode.right
                    elif deleteParent.name < deleteNode.name:
                        deleteParent.right = deleteNode.right
                else:
                    if deleteNode.name < deleteParent.name:
                        deleteParent.left = None
                    else:
                        deleteParent.right = None

            return True
    
        parent = True
        node = self.root

        # find node to remove
        while node and node.name != name:
            parent = node
            if name < node.name:
                node = node.left
            elif name > node.name:
                node = node.right

        # if name does not exst within tree then output an error
        if node is None or node.name != name:
            print("Name does not exist within the PhoneBook\n")
            return False

        # remove node has no children (following BST rules)

        elif node.left is None and node.right is None:
            if name < parent.name:
                parent.left = None
            else:
                parent.right = None
            return True
        
        # remove node has left child only (following BST rules)

        elif node.left and node.right is None:
            if name < parent.name:
                parent.left = node.left
            else:
                parent.right = node.left
            return True

        # remove node that has right child only (following BST rules)

        elif node.left is None and node.right:
            if name < parent.name:
                parent.left = node.right
            else:
                parent.right = node.right
            return True

        # remove node that has left and right children (following BST rules)

        else:
            deleteParent = node
            deleteNode = node.right
            while deleteNode.left:
                deleteParent = deleteNode
                deleteNode = deleteNode.left
            
            node.name = deleteNode.name
            node.phone = deleteNode.phone
            node.address = deleteNode.address
            if deleteNode.right:
                if deleteParent.name > deleteNode.name:
                    deleteParent.left = deleteNode.right
                elif deleteParent.name < deleteParent.name:
                    deleteParent.right = deleteNode.right
            else:
                if deleteNode.name < deleteParent.name:
                    deleteParent.left = None
                else:
                    deleteParent.right = None
            
            return True

    # inorder function to display nodes added in an order of inorder
    def inorder(self):
        return self.rec_inorder(self.root, [])
    # recursion of inorder
    def rec_inorder(self, presentNode, lst):
        if presentNode.left:
            self.rec_inorder(presentNode.left, lst)
        if presentNode:
            lst.append(presentNode)
        if presentNode.right:
            self.rec_inorder(presentNode.right, lst)
        return lst

class Phone:
    def __init__(self):
        # set the root to empty
        self.root = None
    
    # string function to display phone contacts
    def __str__(self):
        node_strs = [str(node) for node in self.inorder()]
        return "-----------------\n".join(node_strs)

    # insert function to insert phone numbers into BST
    def insert(self, name, phone, address):
        if self.root is None:
            self.root = Node(name, phone, address)
        else:
            self.rec_insert(self.root, name, phone, address)
    # recursion of insert function
    def rec_insert(self, presentNode, name, phone, address):
        if phone < presentNode.phone:
            if presentNode.left:
                self.rec_insert(presentNode.left, name, phone, address)
            else:
                presentNode.left = Node(name, phone, address)
        elif phone > presentNode.phone:
            if presentNode.right:
                self.rec_insert(presentNode.right, name, phone, address)
            else:
                presentNode.right = Node(name, phone, address)
    # find function to search for phone numbers within phonebook
    def find(self, phone):
        self.rec_find(phone, self.root)
        if phone is None:
            print(node)
        else:
            return False
    # recursion of find function
    def rec_find(self, phone, node):
        if phone == node.phone:
            print(node)
            return True
        # if phone number does not exist in phonebook print error
        if phone < node.phone:
            if node.left == None:
                print("Phone number does not exist in PhoneBook\n")
                return False
            return self.rec_find(phone, node.left)
        # if phone number does not exist in phonebook print error
        if node.right == None:
            print("Phone number does not exist in PhoneBook\n")
            return False
        return self.rec_find(phone, node.right)
    '''
    remove function in order to remove a phone from the phonebook
    function was created with the aid of https://www.youtube.com/watch?v=LSju119w8BE
    '''
    def remove(self, phone):
        # initaialize root as empty
        if self.root is None:
            return False
        
       # search for name within tree

        elif self.root.phone == phone:
            if self.root.left is None and self.root.right is None:
                self.root == None
            elif self.root and self.root.right is None:
                self.root == self.root.left
            elif self.root.left is None and self.root.right:
                self.root == self.root.right
            elif self.root.left and self.root.right:
                deleteParent = self.root
                deleteNode = self.root.right
                while deleteNode.left:
                    deleteParent = deleteNode
                    deleteNode = deleteNode.left
                
                self.root.name = deleteNode.name
                self.root.phone = deleteNode.phone
                self.root.address = deleteNode.address

                if deleteNode.right:
                    if deleteParent.phone > deleteNode.phone:
                        deleteParent.left = deleteNode.right
                    elif deleteParent.phone < deleteNode.phone:
                        deleteParent.right = deleteNode.right
                else:
                    if deleteNode.phone < deleteParent.phone:
                        deleteParent.left = None
                    else:
                        deleteParent.right = None

            return True
    
        parent = True
        node = self.root

        # find node to remove
        while node and node.phone != phone:
            parent = node
            if phone < node.phone:
                node = node.left
            elif phone > node.phone:
                node = node.right

        # if number does not exst within tree then output an error
        if node is None or node.phone != phone:
            print("Phone number does not exist within the PhoneBook\n")
            return False

        # remove node has no children (following BST rules)

        elif node.left is None and node.right is None:
            if phone < parent.phone:
                parent.left = None
            else:
                parent.right = None
            return True
        
        # remove node has left child only (following BST rules)

        elif node.left and node.right is None:
            if phone < parent.phone:
                parent.left = node.left
            else:
                parent.right = node.left
            return True

        # remove node that has right child only (following BST rules)

        elif node.left is None and node.right:
            if phone < parent.phone:
                parent.left = node.right
            else:
                parent.right = node.right
            return True

        # remove node that has left and right children (following BST rules)

        else:
            deleteParent = node
            deleteNode = node.right
            while deleteNode.left:
                deleteParent = deleteNode
                deleteNode = deleteNode.left
            
            node.phone = deleteNode.phone
            node.name = deleteNode.name
            node.address = deleteNode.address
            if deleteNode.right:
                if deleteParent.phone > deleteNode.phone:
                    deleteParent.left = deleteNode.right
                elif deleteParent.phone < deleteParent.phone:
                    deleteParent.right = deleteNode.right
            else:
                if deleteNode.phone < deleteParent.phone:
                    deleteParent.left = None
                else:
                    deleteParent.right = None
            
            return True
    # inorder function to display nodes added in an order of inorder
    def inorder(self):
        return self.rec_inorder(self.root, [])
    # recursion of inorder  
    def rec_inorder(self, presentNode, lst):
        if presentNode.left:
            self.rec_inorder(presentNode.left, lst)
        if presentNode:
            lst.append(presentNode)
        if presentNode.right:
            self.rec_inorder(presentNode.right, lst)
        return lst

# Main function where functions and operations take place to display outputs and the phonebook

def main():
    # Assigning variables to call classes for name and phone
    phone = Phone()
    name = Name()

    # print function to display a welcoming to phonebook
    print("--------Welcome to the PhoneBook--------\nHere is a list of contacts contained within the PhoneBook:\n")
    # displaying contacts using name insert function
    name.insert("Joao", "08356456", "Portugal")
    name.insert("Cillian", "08746564", "Wexford")
    name.insert("John", "085546456", "Dublin 4")
    name.insert("Terry", "081665464", "Dublin 2")
    name.insert("Conor", "089234324", "Dublin 1")
    name.insert("Aisling", "082324324", "Dublin 13")
    # print contacts
    print(name)
    # displaying contacts using phone insert function
    phone.insert("Adam", "083423432", "Portugal")
    phone.insert("Brendo", "084464647", "Wexford")
    phone.insert("Stacey", "085656456", "Dublin 4")
    phone.insert("Amy", "081645646", "Dublin 2")
    phone.insert("Kerri", "0896456456", "Dublin 1")
    phone.insert("Kanya", "0826456464", "Dublin 13")
    # print contacts
    print(phone)
    
    # use of find function to see if name exits within phonebook
    print("Let's see if Aisling is in the PhoneBook:")
    name.find("Aisling")
    print("Let's see if Timmy is in the PhoneBook:")
    # if name does not exist an error will occur
    name.find("Timmy")

    # use of find function to see if phone exits within phonebook
    print("Let's see if 081645646 is in the PhoneBook:")
    phone.find("081645646")
    print("Let's see if 081645646 is in the PhoneBook:")
    # if phone does not exist an error will occur
    phone.find("0349383863")

    # use of remove function to remove a certain name and its contacts from the phonebook
    print("Remove Joao from PhoneBook:\n")
    print("Updated PhoneBook")
    name.remove("Joao")

    # dislay new updated phonebook
    print(name, phone)
    

    print("Remove Chandler from PhoneBook:")
    # if trying to remove a name and name does not exist an error is outputted
    name.remove("Chandler")
    # use of remove function to remove a certain phone and its contacts from the phonebook
    print("Remove 0896456456 from PhoneBook:\n")
    print("Updated PhoneBook")
    phone.remove("0896456456")

    # dislay new updated phonebook
    print(name, phone)

    print("Remove 09745734 from PhoneBook:")
    # if phone does not exist in phonebook, an error is outputted
    phone.remove("09745734")

if __name__ == "__main__":
    main()

# resources
# https://www.youtube.com/watch?v=LSju119w8BE
# https://www.tutorialspoint.com/python_data_structure/python_tree_traversal_algorithms.htm
# https://qvault.io/python/binary-search-tree-in-python/

File: Assignment 1/bst_phonebook.c
#include <stdio.h>
#include <stdlib.h>

/* TreeNode function to assign values as pointers necessary for further functions
   With aid of // https://stackoverflow.com/questions/33715301/inserting-nodes-in-a-binary-search-tree-c/33715699
   as many errors occured within compiler and this implementation came to be the only reasonable solution.
*/
typedef struct TreeNode {
    struct TreeNode *left;
    struct TreeNode *right;
    char *name;
    char *phone;
    char *address;
} TreeNode;

// newNode function to assign new nodes to the BST. With aid of https://www.log2base2.com/data-structures/tree/insert-a-node-in-binary-search-tree.html.
TreeNode *newNode(char *title, char *mobile, char *dir){

    TreeNode *new_node = (TreeNode *)malloc(sizeof(TreeNode));
    new_node->left = NULL;
    new_node->right = NULL;
    new_node->name = title;
    new_node->phone = mobile;
    new_node->address = dir;
    return new_node;
    
}
// Insert function to insert name into BST. With aid of https://www.log2base2.com/data-structures/tree/insert-a-node-in-binary-search-tree.html
TreeNode* insertName(TreeNode* root, char *title, char *mobile, char *dir){
    if(root == NULL)
        return newNode(title, mobile, dir);
    else if(root->name < title)
        root->right = insertName(root->right, title, mobile, dir);
    else if(root->name > title)
        root->left = insertName(root->left, title, mobile, dir);
    return root;  
}

// Find function to see if name exists within phonebook
TreeNode* findName(TreeNode *root, char *title){
    if(root == NULL) // If it does not exist
        printf("\nName has not been found\n"); // Error case if name does not exist in phonebook
    else if(root->name == title) // If found return success and contacts linked to phone number
        printf("\nName has been found\n - Phone: %s\n - Address: %s\n", root->phone, root->address);
    else if(root->name < title) // BST implementation of rules
        return findName(root->right, title);
    else if(root->name > title)
        return findName(root->right, title);
    else
        return root->name;
}
// Find function to see if phone exists within phonebook
TreeNode* findPhone(TreeNode *root, char *mobile){
    if(root == NULL) //If it does not exist
        printf("\nPhone number has not been found.\n"); // error case if phone number does not exist in phonebook
    else if(root->phone == mobile) // If found return success and contacts linked to phone number
        printf("\nPhone number has been found\n - Name: %s\n - Address: %s\n", root->name, root->address);
    else if(root->phone < mobile) //BST implementation of rules
        return findPhone(root->right, mobile);
    else if(root->phone > mobile)
        return findPhone(root->right, mobile);
    else
        return root;
}
// Function linked to remove functions in order to operate. Minimum function finds the smallest node.
TreeNode* findMinimum(TreeNode *root){
    if(root == NULL)
        return NULL;
    else if(root->right == NULL)
        return root;
    else
        return findMinimum(root->right);
}
 // Remove function for name. Implementation was supported by https://www.codesdope.com/blog/article/binary-search-tree-in-c/
TreeNode* removeName(TreeNode *root, char *title){
    if(root == NULL) // If it does not exist in phonebook display error
        printf("name does not exist within Phonebook\n");
    else if(root->name < title)
        root->right = removeName(root->right, title);
    else if(root->name > title)
        root->left = removeName(root->left, title);
    else
    {
        if(root->left == NULL && root->right == NULL)
        {
            free(root);
            return NULL;
        }
        else if(root->left == NULL || root->right == NULL)
        {
            TreeNode *temp;
            if(root->left == NULL)
                temp = root->right;
            else
                temp = root->left;
            free(root);
            return temp;
        }
        else
        {
            TreeNode *temp = findMinimum(root->right);
            root->name = temp->name;
            root->right = removeName(root->right, temp->name);
        }
    }
    return root;
}
 // Remove function for phone number. Implementation was supported by https://www.codesdope.com/blog/article/binary-search-tree-in-c/
TreeNode* removePhone(TreeNode *root, char *mobile){
    if(root == NULL) // If it does not exist in phonebook display error
        printf("Phone number does not exist within Phonebook\n");
    else if(root->phone < mobile)
        root->right = removePhone(root->right, mobile);
    else if(root->phone > mobile)
        root->left = removePhone(root->left, mobile);
    else
    {
        if(root->left == NULL && root->right == NULL)
        {
            free(root);
            return NULL;
        }
        else if(root->left == NULL || root->right == NULL)
        {
            TreeNode *temp;
            if(root->left == NULL)
                temp = root->right;
            else
                temp = root->left;
            free(root);
            return temp;
        }
        else
        {
            TreeNode *temp = findMinimum(root->right);
            root->phone = temp->phone;
            root->right = removePhone(root->right,
             temp->phone);
        }
    }
    return root;
}
// Inorder function to dislayed roots in an order of Inorder
void inorder(TreeNode *root){
    if(root == NULL) // If nothing display nothing
        return;
    inorder(root->left);
    printf(" | %s - %s - %s\n",
    root->name, 
    root->phone, 
    root->address);
    inorder(root->right);
}
// Main function where phonebook is displayed and where functions are displayed
int main(){
    TreeNode *root;

    printf("--------Welcome to the phonebook--------\nHere is a list of contacts contained within the PhoneBook:\n");

    root = insertName(root, "Leighton Adkins", "0868122244", "Dublin 11");
    root = insertName(root, "Umair Lake", "0869861452", "Dublin 24");
    root = insertName(root, "Heath Mair", "0883418472", "Dublin 4");
    root = insertName(root, "Sofia Stark", "0814672830", "Dublin 1");
    root = insertName(root, "Jean-Luc Wiggins", "0859786732", "Dublin 3");
    root = insertName(root, "Austen Daugherty", "0833336666", "Dublin 2");
    root = insertName(root, "Romario Cox", "0812345678", "Dublin 14");
    root = insertName(root, "Jolyon Sinclair", "0884628422", "Dublin 11");
    root = insertName(root, "Ajay Chamberlain", "0872548194", "Dublin 9");
    root = insertName(root, "Cathy Hood", "0874528573", "Dublin 8");
    root = insertName(root, "Lyra Herrera", "0845389502", "Dublin 6");
    root = insertName(root, "Guy Smart", "0875492641", "Dublin 2");
    root = insertName(root, "Madina Langley", "0857830909", "Dublin 11");
    root = insertName(root, "Zeshan Glover", "0864891432", "Dublin 4");
    root = insertName(root, "Janice Frye", "0854782916", "Dublin 2a");
    root = insertName(root, "Joao Pereira", "0833294067", "Portugal");
    root = insertName(root, "Cillian Damond", "0867563353", "Wexford");
    root = insertName(root, "Aisling Stokes", "0833267892", "Dublin");
    // Inorder function in operation to display phonebook in coorect seuence
    inorder(root);

    // Find function in operation for name
    printf("\nLet's see if Timmy Rodgers is in the phonebook:");
    findName(root, "");
    printf("\nLet's see if Cillian Damond is in the phonebook:");
    findName(root, "Cillian Damond");
    printf("\nLet's see if Aisling Stokes is in the phonebook:");
    findName(root, "Aisling Stokes");

    // Remove function in operation for name
    printf("\nRemove Joao from Phonebook:\n");
    root = removeName(root, "Joao Pereira");

    printf("--Here is the updated Phonebook--\n");
    inorder(root);
 
    // Find function in operation for phone
    printf("\nLet's see if 0833294067 is in the phonebook:");
    findPhone(root, "");
    printf("\nLet's see if 0869861452 is in the phonebook:");
    findPhone(root, "0869861452");
    printf("\nLet's see if 0833267892 is in the phonebook:");
    findPhone(root, "0833267892");

    // Remove function in operationfor phone numbers
    printf("\nRemove 0869861452 from Phonebook:\n");
    printf("--Here is the updated Phonebook--\n");
    root = removePhone(root, "0869861452");

    inorder(root);
    // Remove function in operation if phone number does not exist or is not within phonebook
    printf("\nRemove 050275854 from PhoneBook:\n");
    root = removePhone(root, "");
    printf("--Here is the updated Phonebook--\n");

    inorder(root);

    return 0;
}

// References
// https://www.log2base2.com/data-structures/tree/insert-a-node-in-binary-search-tree.html
// https://www.programiz.com/dsa/binary-search-tree
// https://www.codesdope.com/blog/article/binary-search-tree-in-c/
// https://stackoverflow.com/questions/27285565/c-warning-incompatible-pointer-types-passing
// https://stackoverflow.com/questions/33715301/inserting-nodes-in-a-binary-search-tree-c/33715699
